ci: Move dist-aarch64-linux to an aarch64 runner

mrkajetanp opened the pull request:
Move the dist-aarch64-linux CI job to an aarch64 runner instead of cross-compiling it from an x86 one. This will make it possible to perform optimisations such as LTO, PGO and BOLT later on.

r? @Kobzol

Reland of #133809 now that the higher page sizes are fixed.

try-job: dist-aarch64-linux
try-job: dist-x86_64-linux
try-job: dist-i686-linux

mrkajetanp committed:
ci: Organise shared helper scripts

Move shared helper scripts used by Docker builds under docker/scripts.

mrkajetanp commented:
Same as before, plus a couple small changes to account for the LTO+clang workarounds that were merged in the meantime

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 7d9903c77da3eb88e7f449668ec434e13a710a83 with merge a8e7b2194e6bc485f3cf2d2d78f7d074adb06c36...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7d9903c77da3eb88e7f449668ec434e13a710a83","merge_sha":"a8e7b2194e6bc485f3cf2d2d78f7d074adb06c36"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

mrkajetanp commented:
Do we need a longer timeout here again?

Kobzol commented:
We probably turned that one to a free runner recently-ish, so it times out now. With the cache primed, it should be fine.

@bors try

bors commented:
:hourglass: Trying commit 7d9903c77da3eb88e7f449668ec434e13a710a83 with merge 4c7352d66b905b35c4dd9ea97771e9d2ee241492...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7d9903c77da3eb88e7f449668ec434e13a710a83","merge_sha":"4c7352d66b905b35c4dd9ea97771e9d2ee241492"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #135286) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

mrkajetanp commented:
Ahh I see, I think we need to apply the same fix as in #134690 to i686 as well.

tgross35 commented on src/ci/docker/host-x86_64/dist-aarch64-linux/Dockerfile:
There is the path `src/ci/docker/host-*/disabled` for jobs that are still useful but don't run in CI. I'm not sure what the policy is, but maybe it would be good to move this there instead of removing it? I would find it occasionally useful for verifying aarch64 builds on x86 machines. (non-dist would be better yet, but leaving the dist job around still seems convenient).

Kobzol replied:
Jobs in the `disabled` directory rot fast, and they can be restored from git history, so I don't personally see a lot of value in this. You can always copy-paste the contents locally to test it.

Kobzol commented on src/ci/docker/scripts/build-clang.sh:
```diff
@@ -34,12 +34,12 @@ hide_output \
       -DLLVM_INCLUDE_EXAMPLES=OFF \
       -DLLVM_ENABLE_PROJECTS="clang;lld;compiler-rt;bolt" \
-      -DLLVM_BINUTILS_INCDIR="/rustroot/lib/gcc/x86_64-pc-linux-gnu/$GCC_VERSION/plugin/include/" \
+      -DLLVM_BINUTILS_INCDIR="/rustroot/lib/gcc/$GCC_BUILD_TARGET/$GCC_VERSION/plugin/include/" \
```
It seems to me that this is the only change that happened to the 32-bit runner. I think that the problem is that only the 64-bit version of binutils is installed in the CentOS image:
```
2025-01-09T14:25:36.9324939Z #13 5.216 ---> Package binutils.x86_64 0:2.27-44.base.el7 will be updated
2025-01-09T14:25:36.9325811Z #13 5.217 ---> Package binutils.x86_64 0:2.27-44.base.el7_9.1 will be an update
```
So either this should be reverted (keep it hardcoded for 64-bit), or the 32-bit image should install the 32-bit version of binutils.

mrkajetanp replied:
I'm fairly sure the problem is here
```
-DLLVM_BINUTILS_INCDIR="/rustroot/lib/gcc/$GCC_BUILD_TARGET/$GCC_VERSION/plugin/include/" \
```
Because GCC_BUILD_TARGET will be i686 for the i686 job but the plugin/include will end up under x86_64 because that's where the job is running. I'm running a local build to confirm atm

Kobzol replied:
I don't think that the cause is necessarily that the host is x64, rather than just the binutils package is installed only in the x64 version, not the i686 version. So downloading a binutils.i686 CentOS package should hopefully fix it. If not, I would just hardcode the previous x64 path.

mrkajetanp commented:
ah sorry about that

mrkajetanp committed:
ci: Move dist-aarch64-linux to an aarch64 runner

Move the dist-aarch64-linux CI job to an aarch64 runner instead of
cross-compiling it from an x86 one. This will make it possible to
perform optimisations such as LTO, PGO and BOLT later on.

Kobzol commented:
It seems like you did both, i.e. install 32-bit binutils, but use the x64 path? :D

Anyway, let's see if it helped.

@bors try

bors commented:
:hourglass: Trying commit 307480a15f22a86f0df633ba418352b9a36c6acd with merge c2013d90324bd57e903ea300e181b37f130d4b74...
<!-- homu: {"type":"TryBuildStarted","head_sha":"307480a15f22a86f0df633ba418352b9a36c6acd","merge_sha":"c2013d90324bd57e903ea300e181b37f130d4b74"} -->

mrkajetanp commented:
Well I figured the builds take so long to run that I might as well try the more conservative approach :) I think it needed the i686 binutils it probably would have failed sooner, but it failing on the linking stage of building clang suggests that it's the linker plugin include that was messed up

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12711864691/job/35449700300)
Build commit: c2013d90324bd57e903ea300e181b37f130d4b74 (`c2013d90324bd57e903ea300e181b37f130d4b74`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12711864691/job/35449700300"},"merge_sha":"c2013d90324bd57e903ea300e181b37f130d4b74"} -->

Kobzol commented:
@bors try

(To see cached durations).

bors commented:
:hourglass: Trying commit 307480a15f22a86f0df633ba418352b9a36c6acd with merge 291f58ebd9118530a988bc921530bd05628562ce...
<!-- homu: {"type":"TryBuildStarted","head_sha":"307480a15f22a86f0df633ba418352b9a36c6acd","merge_sha":"291f58ebd9118530a988bc921530bd05628562ce"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12716248075/job/35454350778)
Build commit: 291f58ebd9118530a988bc921530bd05628562ce (`291f58ebd9118530a988bc921530bd05628562ce`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12716248075/job/35454350778"},"merge_sha":"291f58ebd9118530a988bc921530bd05628562ce"} -->

Kobzol commented:
@rust-timer build 291f58ebd9118530a988bc921530bd05628562ce

rust-timer commented:
Queued 291f58ebd9118530a988bc921530bd05628562ce with parent 336209eef13882bd1e211b24779584cb7ef911eb, future [comparison URL](https://perf.rust-lang.org/compare.html?start=336209eef13882bd1e211b24779584cb7ef911eb&end=291f58ebd9118530a988bc921530bd05628562ce).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.7 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (291f58ebd9118530a988bc921530bd05628562ce): [comparison URL](https://perf.rust-lang.org/compare.html?start=336209eef13882bd1e211b24779584cb7ef911eb&end=291f58ebd9118530a988bc921530bd05628562ce&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=336209eef13882bd1e211b24779584cb7ef911eb&end=291f58ebd9118530a988bc921530bd05628562ce&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=336209eef13882bd1e211b24779584cb7ef911eb&end=291f58ebd9118530a988bc921530bd05628562ce&stat=max-rss)
<details>
<summary>Results (primary 2.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6% | [2.6%, 2.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.6% | [2.6%, 2.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=336209eef13882bd1e211b24779584cb7ef911eb&end=291f58ebd9118530a988bc921530bd05628562ce&stat=cycles:u)
<details>
<summary>Results (primary 3.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.4% | [2.6%, 5.5%] | 4     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.4% | [2.6%, 5.5%] | 4     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=336209eef13882bd1e211b24779584cb7ef911eb&end=291f58ebd9118530a988bc921530bd05628562ce&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 763.192s -> 763.291s (0.01%)
**Artifact size**: 325.76 MiB -> 325.62 MiB (-0.04%)

mrkajetanp commented on src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile:
```diff
@@ -0,0 +1,101 @@
+ENV SCRIPT python3 ../x.py dist --host $HOSTS --target $HOSTS
+
+ENV CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER=clang
+ENV DIST_SRC 1
```
no worries, there you go

Kobzol commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc -->
<!-- homu: {"type":"Approved","sha":"9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc with merge 724f36a1d6ede0ebb0a40ff742cae1fbee196c83...
<!-- homu: {"type":"BuildStarted","head_sha":"9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc","merge_sha":"724f36a1d6ede0ebb0a40ff742cae1fbee196c83"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12727471715/job/35477045910) [(plain)](https://github.com/rust-lang-ci/rust/commit/724f36a1d6ede0ebb0a40ff742cae1fbee196c83/checks/35477045910/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12727471715/job/35477827607)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12727471715/job/35477827607","builder_name":"checks-actions"} -->

mrkajetanp commented:
Huh, I don' think this is related to the PR?
```
2025-01-11T21:44:43.5293056Z [1m[31merror[0m[1m:[0m failed to remove file `C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage1-rustc\x86_64-pc-windows-msvc\release\rustc-main.exe`
2025-01-11T21:44:43.5293778Z 
2025-01-11T21:44:43.5293891Z Caused by:
2025-01-11T21:44:43.5294173Z   Access is denied. (os error 5)
```
But then I know nothing about windows, so if someone who does could chime in that'd be great..

Kobzol commented:
Yeah, that's sadly a known flaky failure.

@bors retry

bors commented:
:hourglass: Testing commit 9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc with merge afb3e7ed25d652a9e36da23156faac4184460eb8...
<!-- homu: {"type":"BuildStarted","head_sha":"9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc","merge_sha":"afb3e7ed25d652a9e36da23156faac4184460eb8"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12729265845/job/35480851391) [(plain)](https://github.com/rust-lang-ci/rust/commit/afb3e7ed25d652a9e36da23156faac4184460eb8/checks/35480851391/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12729265845/job/35481808057)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12729265845/job/35481808057","builder_name":"checks-actions"} -->

mrkajetanp commented:
It seems the CI gods are not with us tonight
```
2025-01-12T02:33:21.7595180Z [1m[31merror[0m[1m:[0m failed to download from `https://static.crates.io/crates/rand_xorshift/0.3.0/download`
2025-01-12T02:33:21.7595798Z 
2025-01-12T02:33:21.7595923Z Caused by:
2025-01-12T02:33:21.7596566Z   [7] Could not connect to server (Failed to connect to static.crates.io port 443 after 1019 ms: Could not connect to server)
2025-01-12T02:33:21.7619188Z Build completed unsuccessfully in 0:57:06
```

Kobzol commented:
@bors retry

bors commented:
:hourglass: Testing commit 9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc with merge c0f6a1ce3f89a3240efe5c83839675e20f84e001...
<!-- homu: {"type":"BuildStarted","head_sha":"9ba34f83474ac011db5af6a4d9ef9aa632a6b6fc","merge_sha":"c0f6a1ce3f89a3240efe5c83839675e20f84e001"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12733249512/job/35492088501)
Approved by: Kobzol
Pushing c0f6a1ce3f89a3240efe5c83839675e20f84e001 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Kobzol","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12733249512/job/35492088501"},"merge_sha":"c0f6a1ce3f89a3240efe5c83839675e20f84e001"} -->

rust-timer commented:
Finished benchmarking commit (c0f6a1ce3f89a3240efe5c83839675e20f84e001): [comparison URL](https://perf.rust-lang.org/compare.html?start=1b41e8406b5f9265b414df66f00332b6244a0485&end=c0f6a1ce3f89a3240efe5c83839675e20f84e001&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1b41e8406b5f9265b414df66f00332b6244a0485&end=c0f6a1ce3f89a3240efe5c83839675e20f84e001&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.4%, -0.3%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1b41e8406b5f9265b414df66f00332b6244a0485&end=c0f6a1ce3f89a3240efe5c83839675e20f84e001&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=1b41e8406b5f9265b414df66f00332b6244a0485&end=c0f6a1ce3f89a3240efe5c83839675e20f84e001&stat=cycles:u)
<details>
<summary>Results (secondary -3.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.0%, -3.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=1b41e8406b5f9265b414df66f00332b6244a0485&end=c0f6a1ce3f89a3240efe5c83839675e20f84e001&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 761.975s -> 761.573s (-0.05%)
**Artifact size**: 326.11 MiB -> 326.13 MiB (0.01%)
